src/rust/engine/options/src/cli_alias.rs

```diff
@@ -0,0 +1,119 @@
+// Copyright 2024 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+use crate::Scope;
+use lazy_static::lazy_static;
+use regex::Regex;
+use std::collections::{HashMap, HashSet};
+
+lazy_static! {
+    static ref VALID_ALIAS_RE: Regex = Regex::new(r"^(--)?\w(\w|-)*\w$").unwrap();
+}
+
+fn validate_alias(
+    known_scopes: &HashSet<String>,
+    known_flags: &HashMap<String, HashSet<String>>,
+    alias: &str,
+) -> Result<(), String> {
+    if !VALID_ALIAS_RE.is_match(alias) {
+        return Err(format!(
+            "Invalid alias in `[cli].alias` option: {}. May only contain alphanumerical \
+            letters and the separators `-` and `_`. Flags can be defined using `--`. \
+            A single dash is not allowed.",
+            alias
+        ));
+    }
+
+    if known_scopes.contains(alias) {
+        return Err(format!(
+            "Invalid alias in `[cli].alias` option: {}. This is already a registered goal or subsytem.",
+            alias
+        ));
+    };
+
+    for (scope, args) in known_flags.iter() {
+        if args.contains(alias) {
+            return Err(format!(
+                "Invalid alias in `[cli].alias` option: {}. This is already a registered flag in the {} scope.",
+                alias, Scope::named(scope).name()
+            ));
+        }
+    }
+    Ok(())
+}
+
+fn recursive_expand(
+    definitions: &HashMap<String, Vec<String>>,
+    definition: &Vec<String>,
+    trail: &mut Vec<String>,
+) -> Result<Vec<String>, String> {
+    let mut ret: Vec<String> = vec![];
+    for arg in definition {
+        if let Some(defn) = definitions.get(arg) {
+            trail.push(arg.to_owned());
+            if trail.iter().position(|x| x == arg).unwrap() < trail.len() - 1 {
+                return Err(format!(
+                    "CLI alias cycle detected in `[cli].alias` option:\n{}",
+                    trail.join(" -> ")
+                ));
+            }
+            ret.extend(recursive_expand(definitions, defn, trail)?);
+            trail.pop();
+        } else {
+            ret.push(arg.to_owned());
+        }
+    }
+    Ok(ret)
+}
+
+pub type AliasMap = HashMap<String, Vec<String>>;
+
+#[allow(dead_code)]
+pub fn create_alias_map(
+    known_scopes: &HashSet<String>,
+    known_flags: &HashMap<String, HashSet<String>>,
```

Comment:
Possibly, but I copied this as-is from the Python implementation and I assumed there was a reason it was done this way. I didn't want to start unraveling that yet.

Reply:
Oh yeah, in Python it's just because we pass a map from scope name to ScopeInfo, so we can know if the scope is a subsystem or a goal. But the only reason we need to know that is to finesse the error message, which seems unnecessary. So I turned that into a set of strings, at which point it is just the key set of the other map (I will verify this though).

Reply:
I just realized that the current check (in Python) is not robust enough: It checks that we don't stomp on the fully-qualified flag, e.g., `pants --scope-foo ...`, but it doesn't check shorthand scoped flags, e.g., `pants scope --foo ...`

Reply:
> I didn't want to start unraveling that yet.

Fine by me!

> I just realized that the current check (in Python) is not robust enough: It checks that we don't stomp on the fully-qualified flag, e.g., pants --scope-foo ..., but it doesn't check shorthand scoped flags, e.g., pants scope --foo ...

Sounds like something to consider for follow-up?